File: c_impl/src/primitives/option/option.c
#include "option.h"
#include <assert.h>
#include "../../platform/mem.h"
#include <string.h>
#include "../primitives_context.h"
#include "../../util/hash.h"

// CubsOption cubs_option_init_primitive(CubsValueTag tag, void* optionalValue)
// {  
//     assert(tag != cubsValueTagUserClass && "Use cubs_option_init_user_class for user defined classes");
//     return cubs_option_init_user_class(cubs_primitive_context_for_tag(tag), optionalValue);
// }

CubsOption cubs_option_init(const CubsTypeContext *context, void *optionalValue)
{
    assert(context != NULL);
    if(optionalValue == NULL) {
        const CubsOption option = {._metadata = {0}, .isSome = false, .context = context};
        return option;
    } else {
        CubsOption option = {._metadata = {0}, .isSome = true, .context = context};
        if(context->sizeOfType <= sizeof(option._metadata)) {
            memcpy((void*)&option._metadata, optionalValue, context->sizeOfType);
        } else {
            void* metadataMem = cubs_malloc(context->sizeOfType, _Alignof(size_t));
            memcpy(metadataMem, optionalValue, context->sizeOfType);
            option._metadata[0] = metadataMem;
        }
        return option;
    }
}

void cubs_option_deinit(CubsOption *self)
{
    if(!self->isSome) {
        return;
    }

    if(self->context->destructor.func.externC != NULL) {
        if(self->context->sizeOfType <= sizeof(self->_metadata)) {
            cubs_context_fast_deinit(&self->_metadata, self->context);
        } else {
            cubs_context_fast_deinit(self->_metadata[0], self->context);
            cubs_free(self->_metadata[0], self->context->sizeOfType, _Alignof(size_t));
        }
    }

    memset((void*)self, 0, sizeof(CubsOption));
}

CubsOption cubs_option_clone(const CubsOption *self)
{
    assert(self->context->clone.func.externC != NULL);

    CubsOption out = {._metadata = {0}, .isSome = self->isSome, .context = self->context};
    if(!self->isSome) {
        return out;
    }

    if(self->context->sizeOfType <= sizeof(self->_metadata)) {
        cubs_context_fast_clone((void*)&out._metadata, cubs_option_get(self), self->context);
    } else {
        void* metadataMem = cubs_malloc(self->context->sizeOfType, _Alignof(size_t));
        cubs_context_fast_clone(metadataMem, cubs_option_get(self), self->context);
        out._metadata[0] = metadataMem;
    }
    return out;
}

const void *cubs_option_get(const CubsOption *self)
{
    assert(self->isSome);

    if(self->context->sizeOfType <= sizeof(self->_metadata)) {
        return &self->_metadata;
    } else {
        return self->_metadata[0];
    }
}

void *cubs_option_get_mut(CubsOption *self)
{
    assert(self->isSome);

    if(self->context->sizeOfType <= sizeof(self->_metadata)) {
        return &self->_metadata;
    } else {
        return self->_metadata[0];
    }
}

void cubs_option_take(void *out, CubsOption *self)
{
    assert(self->isSome);

    if(self->context->sizeOfType <= sizeof(self->_metadata)) {
        memcpy(out, &self->_metadata, self->context->sizeOfType);
    } else {
        memcpy(out, self->_metadata[0], self->context->sizeOfType);
        cubs_free(self->_metadata[0], self->context->sizeOfType, _Alignof(size_t));
    }
    memset((void*)self, 0, sizeof(CubsOption));
}

bool cubs_option_eql(const CubsOption *self, const CubsOption *other)
{
    assert(self->context->eql.func.externC == other->context->eql.func.externC);
    assert(self->context->eql.func.externC != NULL);
    assert(self->context->sizeOfType == other->context->sizeOfType);

    if(self->isSome != other->isSome) {
        return false;
    }

    if(self->isSome) {
        return cubs_context_fast_eql(cubs_option_get(self), cubs_option_get(other), self->context);
    } else {
        return true; // both are null
    }
    
}

size_t cubs_option_hash(const CubsOption *self)
{
    assert(self->context->hash.func.externC != NULL);

    if(!self->isSome) {
        return 0;
    }

    const size_t globalHashSeed = cubs_hash_seed();
    const size_t hashed = cubs_context_fast_hash(cubs_option_get(self), self->context);
    return cubs_combine_hash(globalHashSeed, hashed);
}

File: c_impl/src/primitives/option/option.zig
const std = @import("std");
const expect = std.testing.expect;
const script_value = @import("../script_value.zig");
const ValueTag = script_value.ValueTag;
const RawValue = script_value.RawValue;
const CTaggedValue = script_value.CTaggedValue;
const TaggedValue = script_value.TaggedValue;
const String = script_value.String;
const StructContext = script_value.StructContext;

/// Default initialization makes it a none option
pub fn Option(comptime T: type) type {
    return extern struct {
        const Self = @This();
        pub const SCRIPT_SELF_TAG: ValueTag = .option;
        pub const ValueType = T;

        isSome: bool = false,
        _metadata: [4]?*anyopaque = std.mem.zeroes([4]?*anyopaque),
        context: *const StructContext,

        ///
        pub fn init(value: ?T) Self {
            const valueTag = script_value.scriptTypeToTag(T);
            if (value) |v| {
                var mutValue = v;
                const opt = blk: {
                    if (valueTag != .userClass) {
                        break :blk CubsOption.cubs_option_init_primitive(valueTag, &mutValue);
                    } else {
                        break :blk CubsOption.cubs_option_init_user_class(StructContext.auto(T), &mutValue);
                    }
                };
                return @bitCast(opt);
            } else {
                const opt = blk: {
                    if (valueTag != .userClass) {
                        break :blk CubsOption.cubs_option_init_primitive(valueTag, null);
                    } else {
                        break :blk CubsOption.cubs_option_init_user_class(StructContext.auto(T), null);
                    }
                };
                return @bitCast(opt);
            }
        }

        pub fn deinit(self: *Self) void {
            CubsOption.cubs_option_deinit(self.asRawMut());
        }

        pub fn clone(self: *const Self) Self {
            return @bitCast(CubsOption.cubs_option_clone(self.asRaw()));
        }

        pub fn get(self: *const Self) *const T {
            return @ptrCast(@alignCast(CubsOption.cubs_option_get(self.asRaw())));
        }

        pub fn getMut(self: *Self) *T {
            return @ptrCast(@alignCast(CubsOption.cubs_option_get_mut(self.asRawMut())));
        }

        /// Take ownership of the some optional value.
        /// Deinitializes `self` at the same time.
        pub fn take(self: *Self) T {
            var out: T = undefined;
            CubsOption.cubs_option_take(@ptrCast(&out), self.asRawMut());
            return out;
        }

        pub fn eql(self: *const Self, other: *const Self) bool {
            return CubsOption.cubs_option_eql(self.asRaw(), other.asRaw());
        }

        pub fn hash(self: *const Self) usize {
            return CubsOption.cubs_option_hash(self.asRaw());
        }

        pub fn asRaw(self: *const Self) *const CubsOption {
            return @ptrCast(self);
        }

        pub fn asRawMut(self: *Self) *CubsOption {
            return @ptrCast(self);
        }
    };
}

pub const CubsOption = extern struct {
    isSome: bool = false,
    _metadata: [4]?*anyopaque = std.mem.zeroes([4]?*anyopaque),
    context: *const StructContext,

    const Self = @This();
    pub const SCRIPT_SELF_TAG: ValueTag = .option;

    pub extern fn cubs_option_init_primitive(tag: ValueTag, optionalValue: ?*anyopaque) callconv(.C) Self;
    pub extern fn cubs_option_init_user_class(context: *const StructContext, optionalValue: ?*anyopaque) callconv(.C) Self;
    pub extern fn cubs_option_deinit(self: *Self) callconv(.C) void;
    pub extern fn cubs_option_clone(self: *const Self) callconv(.C) Self;
    pub extern fn cubs_option_get(self: *const Self) callconv(.C) *const anyopaque;
    pub extern fn cubs_option_get_mut(self: *Self) callconv(.C) *anyopaque;
    pub extern fn cubs_option_take(out: *anyopaque, self: *Self) callconv(.C) void;
    pub extern fn cubs_option_eql(self: *const Self, other: *const Self) callconv(.C) bool;
    pub extern fn cubs_option_hash(self: *const Self) callconv(.C) usize;
};

test "null" {
    {
        var opt = Option(i64).init(null);
        defer opt.deinit();

        try expect(opt.isSome == false);
    }
    {
        var opt = Option(String).init(null);
        defer opt.deinit();

        try expect(opt.isSome == false);
    }
    {
        var opt = Option(Option(String)).init(null);
        defer opt.deinit();

        try expect(opt.isSome == false);
    }
}

test "some" {
    {
        var opt = Option(i64).init(4);
        defer opt.deinit();

        try expect(opt.isSome);

        try expect(opt.get().* == 4);
        try expect(opt.getMut().* == 4);

        opt.getMut().* = 5;

        try expect(opt.get().* == 5);
        try expect(opt.getMut().* == 5);
    }
}

test "take" {
    {
        var opt = Option(i64).init(4);
        defer opt.deinit();

        try expect(opt.isSome);

        const took = opt.take();
        try expect(took == 4);
        try expect(opt.isSome == false);
    }
}

test "clone" {
    {
        var nullOpt = Option(i64).init(null);
        defer nullOpt.deinit();

        var clone = nullOpt.clone();
        defer clone.deinit();

        try expect(!clone.isSome);
    }
    {
        var someOpt = Option(i64).init(1);
        defer someOpt.deinit();

        var clone = someOpt.clone();
        defer clone.deinit();

        try expect(clone.isSome);
        try expect(clone.get().* == 1);
    }
}

test "eql" {
    {
        var nullOpt1 = Option(i64).init(null);
        defer nullOpt1.deinit();

        var nullOpt2 = Option(i64).init(null);
        defer nullOpt2.deinit();

        try expect(nullOpt1.eql(&nullOpt2));
    }
    {
        var nullOpt = Option(i64).init(null);
        defer nullOpt.deinit();

        var someOpt = Option(i64).init(1);
        defer someOpt.deinit();

        try expect(!nullOpt.eql(&someOpt));
        try expect(!someOpt.eql(&nullOpt));
    }
    {
        var someOpt1 = Option(i64).init(1);
        defer someOpt1.deinit();

        var someOpt2 = Option(i64).init(1);
        defer someOpt2.deinit();

        try expect(someOpt1.eql(&someOpt2));
    }
    {
        var someOpt1 = Option(i64).init(1);
        defer someOpt1.deinit();

        var someOpt2 = Option(i64).init(2);
        defer someOpt2.deinit();

        try expect(!someOpt1.eql(&someOpt2));
        try expect(!someOpt2.eql(&someOpt1));
    }
}

test "hash" {
    {
        var nullOpt1 = Option(i64).init(null);
        defer nullOpt1.deinit();

        var nullOpt2 = Option(i64).init(null);
        defer nullOpt2.deinit();

        const h1 = nullOpt1.hash();
        const h2 = nullOpt2.hash();

        try expect(h1 == h2);
    }
    {
        var nullOpt = Option(i64).init(null);
        defer nullOpt.deinit();

        var someOpt = Option(i64).init(1);
        defer someOpt.deinit();

        const h1 = nullOpt.hash();
        const h2 = someOpt.hash();

        if (h1 == h2) {
            return error.SkipZigTest;
        }
    }
    {
        var someOpt1 = Option(i64).init(1);
        defer someOpt1.deinit();

        var someOpt2 = Option(i64).init(1);
        defer someOpt2.deinit();

        const h1 = someOpt1.hash();
        const h2 = someOpt2.hash();

        try expect(h1 == h2);
    }
    {
        var someOpt1 = Option(i64).init(1);
        defer someOpt1.deinit();

        var someOpt2 = Option(i64).init(2);
        defer someOpt2.deinit();

        const h1 = someOpt1.hash();
        const h2 = someOpt2.hash();

        if (h1 == h2) {
            return error.SkipZigTest;
        }
    }
}
